### PR TITLE
Update readme file with more project details. 

### PR DESCRIPTION
**Pull request by weiweishi**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,3 +1,51 @@
-# DIDockerImages
-This is repo of Dockerfiles
-to build various usefull images
+# Digital Initiatives Docker Images
+This is the repository to keep track of the Docker project, and all the docker images created for Digital Initiatives projects.
+
+### Project Objective ###
+
+As our DI projects grew in complexity, we have some additional needs for our development/demo/testing environments beyond the underutilized vagrant boxes:
+* lightweight. Vagrant boxes have a huge visualization overhead and are more resource-extensive.
```
Comment (pbinkley) [inline README.md]: Should be "virtualization overhead"

File: README.md [hunk 2]
```diff
@@ -1,3 +1,51 @@
...
+* easy setup for onboarding new developers.
+
+The objectives of the DI Docker projects are:
+* create Dockerfiles (docker-compose files if needed), and Docker images for DI application: HN, HN2, Hyku, Avalon, and DMP for
+  * using as a development environment
+  * using as a demo environment during the development cycle, preferably as a result of CI testing/auto built.
+  * using for one-off sandbox testing and investigation as needed. 
+
+### Requirements / Deliverables  ###
+* Dockerfiles(docker-compose file if needed), and Docker images for DI projects include HN, HN, Avalon, and DMP.
```
Comment (pbinkley) [inline README.md]: Should be "HN, HN2" - and do we need Hyku here (as above in line 14)? I suppose it has some ongoing value in helping us track Hyrax work, but I don't know how long we need to support it as a separate Docker project.

File: README.md [hunk 3]
```diff
@@ -1,3 +1,51 @@
...
+  * using as a demo environment during the development cycle, preferably as a result of CI testing/auto built.
+  * using for one-off sandbox testing and investigation as needed. 
+
+### Requirements / Deliverables  ###
+* Dockerfiles(docker-compose file if needed), and Docker images for DI projects include HN, HN, Avalon, and DMP. 
+..The docker images should
+  * include development and test environment in the development version
+  * run production environment in the demo/sandbox version
+  * whenever possible, be built with the system requirements that are of the same version as in the production systems (e.g. CentoOS, and  Ruby, etc.). 
+  * not depend on internal UAL RPM repos to allow external collaborators to use images
```
Comment (pbinkley) [inline README.md]: Clarify: "... RPM repos, so that external collaborators can use the images"

File: README.md [hunk 4]
```diff
@@ -1,3 +1,51 @@
...
+..The dockerfiles (and docker-compose files if needed) should:
+  * be as similar as possible among projects for long-term maintainability
+  * facilitate all the requirements for the DI docker images
```
Comment (pbinkley) [inline README.md]: Add "* reuse component images such as Fedora, Solr, MySQL as far as possible, to avoid duplication of effort"

**Comment by weiweishi** [comment]
@pbinkley made changes based on your suggestions.

File: README.md [hunk 5]
```diff
@@ -1,3 +1,54 @@
...
+* Dockerfiles(docker-compose files if needed), and Docker images for DI projects include HN, HN2, Avalon, and DMP. 
+..The docker images should
+  * include development and test environment in the development version
+  * run production environment in the demo/sandbox version
+  * whenever possible, be built with the system requirements that are of the same version as in the production systems (e.g. CentoOS, and  Ruby, etc.). 
+  * not depend on packages in internal UAL RPM repos, so that external collaborators can use images. 
+..The dockerfiles (and docker-compose files if needed) should:
+  * be as similar as possible among projects for long-term maintainability
+  * facilitate all the requirements for the DI docker images
+  * reuse component images such as Fedora, Solr, MySQL as far as possible, to avoid duplication of effort.
```
Comment (mbarnett) [inline README.md]: Maybe also add a note that this should ease long-term maintenance versus the original Vagrant files, which had a tendency to break/fall out of date and require a lot of work on our part to keep happy

File: README.md [hunk 6]
```diff
@@ -1,3 +1,54 @@
...
+  * be as similar as possible among projects for long-term maintainability
+  * facilitate all the requirements for the DI docker images
+  * reuse component images such as Fedora, Solr, MySQL as far as possible, to avoid duplication of effort. 
+  * follow community conventions and techniques whenever possible. 
+The project should also include clear documentation:
+* clear in-line documentation within the dockerfiles/docker-compose files
+* README documentation for detailed and well-tested instructions. 
+* Documentation on practices to maintain existing dockerfiles, and create new ones. 
+As needs rise, sandboxing-like dockerfiles and images might be required to track community development work, such as Hyku and DMP Roadmap. These sub-projects should:
+* reuse existing components/dockerfiles as far as possible
```
Comment (mbarnett) [inline README.md]: Similar note about long-term ease of maintenance

File: README.md [hunk 7]
```diff
@@ -1,3 +1,54 @@
...
+This is the repository to keep track of the Docker project, and all the docker images created for Digital Initiatives projects.
+
+### Project Objective ###
+
+As our DI projects grew in complexity, we have some additional needs for our development/demo/testing environments beyond the underutilized vagrant boxes:
+* lightweight. Vagrant boxes have a huge virtualization overhead and are more resource-extensive.
+* performance improvement.
+* easy mapping to the local file system for development.
+* stable platform that can be deployable as a testing/demoing environment as part of the CI. 
+* easy setup for onboarding new developers.
```
Comment (mbarnett) [inline README.md]: Maybe add something about reducing the long-term maintenance burden versus the old Vagrant situation as a top level goal?

**Comment by weiweishi** [comment]
@mbarnett updated the document based on your review.

File: README.md [hunk 8]
```diff
@@ -1,3 +1,56 @@
...
+* create Dockerfiles (docker-compose files if needed), and Docker images for DI application: HN, HN2, Avalon, and DMP for
+  * using as a development environment
+  * using as a demo environment during the development cycle, preferably as a result of CI testing/auto built.
+  * using for one-off sandbox testing and investigation as needed. 
+
+### Requirements / Deliverables  ###
+* Dockerfiles(docker-compose files if needed), and Docker images for DI projects include HN, HN2, Avalon, and DMP. 
+..The docker images should
+  * include development and test environment in the development version
+  * run production environment in the demo/sandbox version
```
Comment (murny) [inline README.md]: Doesn't have to happen here or be documented in this README, but there seems to be a lot of confusion about how this looks like. Maybe I'lll throw all this info below into a issue in this repo?

Currently we are forcing a docker-compose.yml to work on project atomic from my understanding? I would say this  is not the proper way of doing this. There is many ways to tackle this. Project atomic is one (based on kubernetes which is another solution). As far as I know, docker-compose does not work with project atomic. Project atomic expects a different file format and configuration called `Nulecule` (http://www.projectatomic.io/docs/nulecule/) which is similar to docker-compose but not the same.  An example repo of a rails project (gitlab) on project atomic: https://github.com/projectatomic/nulecule-library/tree/master/gitlab-centos7-atomicapp (Look at it's nulecule file here: https://github.com/projectatomic/nulecule-library/blob/master/gitlab-centos7-atomicapp/Nulecule). Which leaves me with some questions of how we are currently doing this.

For how to use docker-compose in production the official solution is to use something called docker swarm. There is a ton of official documentation on this for example here: https://docs.docker.com/compose/production/

Or even the official beginner tutorials will explain this in further details with a good example from start to production: https://docs.docker.com/get-started/
Docker labs also has good documentation: https://github.com/docker/labs
Especially here: https://github.com/docker/labs/tree/master/beginner (python app but very similar to a rails app)

You will see this is quite different then anything we are doing currently. The main difference you will see is their docker-compose.yml has alot more configuration for production (different services, different port assignments, different volumes and environment vars, restart policies, `deploy:` key for scaling/etc)

File: README.md [hunk 9]
```diff
@@ -1,3 +1,56 @@
...
+  * be as similar as possible among projects, have clear documentation and use shared components/components from community work whenever possible for long-term maintainability
+The project should also include clear documentation:
+* clear in-line documentation within the dockerfiles/docker-compose files
+* README documentation for detailed and well-tested instructions. 
+* Documentation on practices to maintain existing dockerfiles, and create new ones. 
+As needs rise, sandboxing-like dockerfiles and images might be required to track community development work, such as Hyku and DMP Roadmap. These sub-projects should:
+* reuse existing components/dockerfiles as far as possible
+* leverage on existing projects and community work as much as possible for long term maintainability. 
+* run on production environment in the demo/sandbox version
+* include development and test environment in the development version
```
Comment (murny) [inline README.md]: this bullet is duplicated. Also found on line 23

File: README.md [hunk 10]
```diff
@@ -1,3 +1,56 @@
...
+* easy setup for onboarding new developers.
+* easy to maintain in the long-term.
+
+The objectives of the DI Docker projects are:
+* create Dockerfiles (docker-compose files if needed), and Docker images for DI application: HN, HN2, Avalon, and DMP for
+  * using as a development environment
+  * using as a demo environment during the development cycle, preferably as a result of CI testing/auto built.
+  * using for one-off sandbox testing and investigation as needed. 
+
+### Requirements / Deliverables  ###
```
Comment (murny) [inline README.md]: Maybe add some CI for custom docker images we create? We agree to CI everything else here, why not docker images too?

Can be done with travisCI quite easily. Lots of examples from official docker repos (https://github.com/docker-library). For example:  https://github.com/docker-library/rails/blob/master/.travis.yml

Looks like the common CI builds for the official docker repos is to download a test script (looks like it runs some metrics/lints on your docker stuff) and simply just build the docker images. If all succeeds, its a success.

File: README.md (hunk 8 above)
Comment (weiweishi) [inline README.md]: I will put in an issue as a spike to investigate Docker Swarm. In our CI document, we mentioned using Atomic as a possible solution for the platform to support our on-demand sandbox requirements. But I'll remove it if we don't think Atomic is the right choice.

File: README.md (hunk 9 above)
Comment (weiweishi) [inline README.md]: I think I messed up my markdowns, this is supposed to be the requirement for Hyku or other docker images that track community development work mentioned on Line 36. I will fix it.

Comment (murny) [inline README.md]: Oh i could be wrong. Ones a sub bullet and others a main bullet. Just viewed the rendered markdown. Nevermind.

**Pull request merged**